pkg/compactor/block_upload.go

```diff
@@ -394,6 +412,98 @@ func (c *MultitenantCompactor) uploadMeta(ctx context.Context, logger log.Logger
 	return nil
 }
 
+func (c *MultitenantCompactor) validateBlock(ctx context.Context, w http.ResponseWriter, blockID ulid.ULID,
+	userBkt objstore.Bucket, meta metadata.Meta) error {
+	blockDir, err := os.MkdirTemp(filepath.Join(c.compactorCfg.DataDir), "upload")
+	if err != nil {
+		return errors.Wrap(err, "failed to create temp dir")
+	}
+	defer func() {
+		if err := os.RemoveAll(blockDir); err != nil {
+			level.Warn(c.logger).Log("msg", "failed to remove temp dir", "path", blockDir, "err", err)
+		}
+	}()
+
+	rps := newRequestProgressSender(w)
+	var objectCount, objectBytes int64
+	if err := userBkt.Iter(ctx, blockID.String(), func(pth string) error {
+		fname := filepath.Join(blockDir, pth)
+		if pth == uploadingMetaFilename {
+			fname = filepath.Join(blockDir, block.MetaFilename) // rename to final meta.json in temp dir
+		}
+
+		r, err := userBkt.Get(ctx, pth)
+		if err != nil {
+			return errors.Wrap(err, fmt.Sprintf("failed to get object %q from object storage", pth))
+		}
+
+		f, err := os.Create(fname)
+		if err != nil {
+			return errors.Wrap(err, "failed creating temp file")
+		}
+		defer func() {
+			_ = f.Close()
+		}()
+		bytesWritten, err := io.Copy(f, r)
+		if err != nil {
+			return err
+		}
+		objectCount++
+		objectBytes += bytesWritten
+		if err := f.Close(); err != nil {
+			return errors.Wrap(err, "failed to close temp file")
+		}
+		if err = rps.SendProgressMessageWithObjectStats("block download in progress", objectCount, objectBytes); err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return errors.Wrapf(err, "failed to iterate block %s", blockID.String())
+	}
+	if err = rps.SendProgressMessage("block download complete"); err != nil {
+		return err
+	}
+
+	// Read metadata file, populate mop of file paths and sizes
+	var blockMetadata metadata.Meta
+	fileStats := map[string]int64{}
+	file, err := ioutil.ReadFile(filepath.Join(blockDir, block.MetaFilename))
+	if err != nil {
+		return errors.Wrap(err, "error reading block metadata")
+	}
+	if err = json.Unmarshal(file, &blockMetadata); err != nil {
+		return errors.Wrap(err, "error reading block metadata json")
+	}
+	// populate relative path to file size map
+	for _, f := range blockMetadata.Thanos.Files {
+		fileStats[f.RelPath] = f.SizeBytes
+	}
+
+	// Check against downloaded names and sizes. Sizes are determined via OS level "stat" command
+	blockFiles, err := block.GatherFileStats(blockDir, metadata.NoneFunc, c.logger)
+	if err != nil {
+		return errors.Wrap(err, "error gathering block file stats")
+	}
+	for _, f := range blockFiles {
+		if f.RelPath == block.MetaFilename {
+			continue
+		}
+		if val, ok := fileStats[f.RelPath]; ok {
+			if val != f.SizeBytes {
+				return errors.Wrap(err, fmt.Sprintf("file size mismatch for %s", f.RelPath))
+			}
+		} else {
+			return errors.Wrap(err, fmt.Sprintf("file not in metadata %s", f.RelPath))
+		}
+	}
```

Comment:
Instead of creating `fileStats`, what if we simply iterate over `blockMetadata.Thanos.Files` and verify that file exists locally and has the right size?

Something like:
```
	for _, f := range blockMetadata.Thanos.Files {
		fi, err := os.Stat(filepath.Join(blockDir, filepath.FromSlash(f.RelPath)))
		if err != nil {
			return errors.Wrapf(err, "failed to stat %s", f.RelPath)
		}

		if !fi.Mode().IsRegular() {
			return errors.Errorf("not a file: %s", f.RelPath)
		}

		if f.RelPath != block.MetaFilename && fi.Size() != f.SizeBytes {
			return errors.Errorf("file size mismatch for %s", f.RelPath)
		}
	}
```

This wouldn't verify that all files from meta.json are there, but wouldn't check if any extra files were uploaded. We could check for extra files in `UploadFile` instead. WDYT?